How many late days are you using for this assignment? 0

In [2]:
# Imports and pip installations (if needed)
import pandas as pd   
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [3]:
# Load the given datasets
categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print(categorical)
print(numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

Explaing Data: 
| Variable | Explanation                                                                 | Feature/Label | Numerical/Categorical |
|----------|-----------------------------------------------------------------------------|---------------|-----------------------|
| age      | Refers to the age of the patient                                            | Feature       | Numerical             |
| bp       | Measure of patient's blood pressure in millimeters of mercury               | Feature       | Numerical             |
| sg       | Refers to specific gravity that reflects the concentration of urine         | Feature       | Categorical           |
| al       | Assesses the level of albumin in urine with levels ranging from 0 to 5     | Feature       | Categorical           |
| su       | Checks presence of sugar in urine, with levels ranging from 0 to 5         | Feature       | Categorical           |
| rbc      | Indicates if red blood cells are normal or abnormal in the urine            | Feature       | Categorical           |
| pc       | It is the presence of pus cells in urine, classified as normal or abnormal  | Feature       | Categorical           |
| pcc      | Shows whether pus cell clumps are present or not                            | Feature       | Categorical           |
| ba       | Indicates the presence of bacteria in the urine                             | Feature       | Categorical           |
| bgr      | Measures random blood glucose levels in mg/dl                               | Feature       | Numerical             |
| bu       | Measurement of blood urea in mg/dl                                          | Feature       | Numerical             |
| sc       | Level of creatinine in serum (blood) in mg/dl                              | Feature       | Numerical             |
| sod      | Measurement of sodium levels in mEq/L                                       | Feature       | Numerical             |
| pot      | Measurement of potassium levels in mEq/L                                     | Feature       | Numerical             |
| hemo     | Indicates the amount of hemoglobin in grams                                   | Feature       | Numerical             |
| pcv      | Shows the proportion of blood volume occupied by red blood cells            | Feature       | Numerical             |
| wc       | The number of white blood cells in cells per cubic millimeter               | Feature       | Numerical             |
| rc       | The number of red blood cells in cells per cubic millimeter                 | Feature       | Numerical             |
| htn      | Indicates the presence (yes) or absence (no) of hypertension                | Feature       | Categorical           |
| dm       | Checks if the patient has diabetes (yes) or does not have it (no)          | Feature       | Categorical           |
| cad      | Indicates the presence (yes) or absence (no) of coronary artery disease     | Feature       | Categorical           |
| appet    | Checks whether the patient has good or poor appetite                        | Feature       | Categorical           |
| pe       | Indicates the presence (yes) or absence (no) of swelling in the feet        | Feature       | Categorical           |
| ane      | Checks whether the patient has anemia or not                                 | Feature       | Categorical           |
| class    | It is a label or target variable that shows whether the patient has CKD     | Label         | Categorical           |



# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [6]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates_numerical = numerical.duplicated()
total_duplicates_numerical = sum(duplicates_numerical)
if(total_duplicates_numerical>0):
    print("Number of duplicate rows in numerical:",total_duplicates_numerical)
duplicate_rows_numerical = numerical[numerical.duplicated(keep=False)]
print("Duplicate rows in numerical:-")
print(duplicate_rows_numerical)

# Drop these duplicate rows
numerical_without_duplicates = numerical.drop_duplicates()
# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates_categorical = categorical.duplicated()
total_duplicates_categorical = sum(duplicates_categorical)
if(total_duplicates_categorical>0):
    print("Number of duplicate rows in categorical:",total_duplicates_categorical)
duplicate_rows_categorical = categorical[categorical.duplicated(keep=False)]
print(duplicate_rows_categorical)
categorical_without_duplicates = categorical.drop_duplicates()


Number of duplicate rows in numerical: 30
Duplicate rows in numerical:-
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
15      224481  54.0   60.0  125.0   21.0  1.3  137.0  3.4  15.0  46.0   
35      431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
42      441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
53      261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
54      721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
61      872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
78      501321   4.0    NaN   99.0   23.0  0.6  138.0  4.4  12.0  34.0   
79      283427  55.0   80.0  140.0   49.0  0.5  150.0  4.9  15.7  47.0   
81      302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
98      708255   3.0    NaN    NaN   22.

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

No, all unique ids are not present in both the data sets. The code below shows that. The rows that are missing from one dataset will be in the merged dataset but will have values NaN for the dataset columns that does not have that specific unique id. The missing unique ids from the numerical dataset are also given after merging the datasets.

In [7]:
unique_ids_df1 = set(categorical_without_duplicates['unique_id'])
unique_ids_df2 = set(numerical_without_duplicates['unique_id'])

# Check for presence of unique IDs in both datasets
missing_in_df1 = unique_ids_df2 - unique_ids_df1
missing_in_df2 = unique_ids_df1 - unique_ids_df2

# Print missing IDs
print("IDs missing from categorical dataset:", missing_in_df1)
print("IDs missing from numerical dataset:", missing_in_df2)

IDs missing from categorical dataset: set()
IDs missing from numerical dataset: {208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}


In [8]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_dataset = pd.merge(numerical_without_duplicates, categorical_without_duplicates, on='unique_id', how='outer', indicator=True)
#Print the combined dataset
print(merged_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

            pcc          ba  htn   dm  cad appet   

In [9]:
missing_rows = merged_dataset[merged_dataset['_merge'] != 'both']
print("\nRows missing from one of the datasets:")
print(missing_rows)


Rows missing from one of the datasets:
     unique_id  age  bp  bgr  bu  sc  sod  pot  hemo  pcv  ...         pcc  \
47      208034  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
58      240430  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...         NaN   
64      254775  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
69      266335  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
80      292312  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
83      298297  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
97      325238  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
173     488947  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
261     663387  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   
376     948317  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  notpresent   

             ba  htn   dm  cad appet   pe  ane  Target      _merge  
47   notpresent  yes  yes  yes  

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [10]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = len(merged_dataset)
missing_values = merged_dataset.isna().any(axis=1).sum()
per_missing = (missing_values/total_rows)*100
# Print %
print(per_missing)
# Drop these rows from the dataset
merged_dataset_withoutna = merged_dataset.dropna()
# Print the Dataset
print(merged_dataset_withoutna)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

            pcc          ba  htn 

## Part 3.4: Sort the dataset according to the Labels

In [12]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_merged_dataset = merged_dataset_withoutna.sort_values(by='Target')
sorted_merged = sorted_merged_dataset.reset_index(drop=True)
# Print the dataset
print(sorted_merged)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

            pcc          ba  htn   dm  cad appet   

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [13]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummy_categorical = pd.get_dummies(sorted_merged, columns=categorical_columns, dtype=int)
# Print the dataset
print(dummy_categorical)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor pe_no 

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

I think that having one single column for binary values in a downstream modeling would be equivalent as it will properly give the information if something is present/absent, yes/no, true/false. If a categorical values like "type of condiment" can take 5 separate values, then we will have 5 separate binary encoded columns for each condiment where 1 will indicate of the particular condiment is present and 0 will indicate its absence. This will help in the prediction of the rating of a particular sandwich by showing which condiments go best together according to my understanding.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [17]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
dataset_removed_outliers = dummy_categorical.copy()
for col in numerical_columns:
    mean = dummy_categorical[col].mean()
    stdev = dummy_categorical[col].std()

    ul = mean+(3*stdev)
    ll = mean-(3*stdev)

    dataset_removed_outliers = dataset_removed_outliers[(dataset_removed_outliers[col]>=ll)&(dataset_removed_outliers[col]<=ul)].reset_index(drop=True)
# Print the dataset
print(dataset_removed_outliers)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
129     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
130     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
131     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
132     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
133     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor pe_no 

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [18]:
# Normalize the all Numerical Attributes in the dataset.
scalar = MinMaxScaler()
dataset_normalized = dataset_removed_outliers.copy()
dataset_normalized[numerical_columns] = scalar.fit_transform(dataset_normalized[numerical_columns])
# Print the dataset
print(dataset_normalized)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
129     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
130     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
131     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
132     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
133     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.657143  0.172131  0.210526  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The unique_id column and the redundant categorical columns are not appropriate for modeling any predictions. A unique_id does not indicate anything related to the disease, its just an identification for a patient but for data analysis we just require the data to draw conclusions by figuring out patterns and what the data indicates. Unique id has no role to play in this. And the binary encoded categorical columns already give the yes/no, presence/absence indication through just a single column so having the exact opposite column is just redundant.

In [19]:
#Remove that column
unnecessary  = ['unique_id', 'rbc_abnormal', 'pc_abnormal', 'pcc_notpresent', 'ba_notpresent', 'htn_no', 'dm_no', 'cad_no', 'appet_poor', 'pe_no', 'ane_no', 'Target_notckd']
dataset_final = dataset_normalized.copy()
dataset_final = dataset_final.drop(columns=unnecessary)
# Print the dataset
print(dataset_final)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
..        ...   ...       ...       ...       ...       ...       ...   
129  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.257143   
130  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.400000   
131  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.371429   
132  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.171429   
133  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.514286   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
0    0.172131  0.210526  0.395161  ...       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [20]:
# Export the dataframe to a new csv file
dataset_final.to_csv('dataset_cvs.csv', index=False)
# Export the dataframe to a new json file
dataset_final.to_json('dataset_json.json', orient='records', index=False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

In [21]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
# If you want to remove the limit on the width of columns
pd.set_option('display.max_colwidth', None)
print(dataset_final.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
5   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.028571   
6   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.542857   
7   0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.285714   
8   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286   
9   0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.514286   
10  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.571429   
11  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.714286   
12  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.57

### Paste the markdown table here


|    age   |   bp   |    bgr    |    bu    |    sc    |    sod   |    pot   |    hemo   |   pcv   |   wbcc   |   rbcc   |   al   |   su   | rbc_normal | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_good | pe_yes | ane_yes | Target_ckd |
|----------|--------|-----------|----------|----------|----------|----------|-----------|---------|----------|----------|--------|--------|------------|-----------|-------------|------------|----------|--------|---------|------------|--------|---------|------------|
| 0.743243 |  0.50  | 0.442060  | 0.901961 | 0.432099 | 0.500000 | 0.657143 | 0.172131  | 0.210526 | 0.395161 | 0.153846 |  2.0   |  0.0   |     0      |     0     |      0      |     0      |     1    |    1   |    1    |      0     |   1    |    1    |     1      |
| 0.770270 |  1.00  | 0.901288  | 0.163399 | 0.345679 | 0.766667 | 0.171429 | 0.606557  | 0.631579 | 0.443548 | 0.410256 |  2.0   |  2.0   |     1      |     1     |      0      |     1      |     1    |    0   |    1    |      1     |   0    |    0    |     1      |
| 0.905405 |  0.50  | 0.785408  | 0.862745 | 0.518519 | 0.600000 | 0.828571 | 0.401639  | 0.447368 | 0.233871 | 0.435897 |  2.0   |  0.0   |     0      |     0     |      0      |     0      |     1    |    1   |    1    |      1     |   0    |    0    |     1      |
| 0.202703 |  0.75  | 0.158798  | 0.196078 | 0.160494 | 0.166667 | 0.171429 | 0.221311  | 0.184211 | 0.653226 | 0.333333 |  4.0   |  0.0   |     1      |     0     |      1      |     1      |     0    |    0   |    0    |      1     |   0    |    1    |     1      |
| 0.783784 |  1.00  | 0.399142  | 0.287582 | 0.839506 | 0.666667 | 0.485714 | 0.188525  | 0.263158 | 0.258065 | 0.205128 |  4.0   |  2.0   |     0      |     0     |      0      |     1      |     1    |    1   |    0    |      1     |   1    |    0    |     1      |
| 0.729730 |  0.00  | 0.935622  | 0.169935 | 0.160494 | 0.333333 | 0.028571 | 0.188525  | 0.236842 | 0.879032 | 0.102564 |  3.0   |  1.0   |     1      |     0     |      1      |     1      |     1    |    0   |    0    |      0     |   0    |    0    |     1      |
| 0.675676 |  0.75  | 0.253219  | 0.633987 | 0.777778 | 0.366667 | 0.542857 | 0.286885  | 0.342105 | 0.169355 | 0.205128 |  2.0   |  0.0   |     0      |     0     |      0      |     0      |     1    |    0   |    0    |      1     |   0    |    0    |     1      |
| 0.851351 |  0.25  | 0.618026  | 0.562092 | 0.728395 | 0.000000 | 0.285714 | 0.311475  | 0.315789 | 0.580645 | 0.179487 |  4.0   |  3.0   |     1      |     0     |      1      |     1      |     1    |    1   |    1    |      1     |   1    |    1    |     1      |
| 0.540541 |  0.00  | 0.399142  | 0.535948 | 0.358025 | 0.700000 | 0.314286 | 0.344262  | 0.315789 | 0.830645 | 0.153846 |  1.0   |  0.0   |     1      |     1     |      0      |     0      |     1    |    1   |    0    |      1     |   0    |    0    |     1      |
| 0.756757 |  0.25  | 0.223176  | 0.209150 | 0.160494 | 0.533333 | 0.514286 | 0.573770  | 0.605263 | 0.290323 | 0.333333 |  3.0   |  0.0   |     1      |     0     |      0      |     0      |     1    |    1   |    0    |      1     |   0    |    0    |     1      |
| 0.662162 |  0.50  | 0.618026  | 0.411765 | 0.432099 | 0.566667 | 0.571429 | 0.434426  | 0.473684 | 0.250000 | 0.282051 |  3.0   |  1.0   |     1      |     0     |      1      |     1      |     1    |    1   |    0    |      1     |   1    |    0    |     1      |
| 0.783784 |  0.00  | 0.725322  | 0.313725 | 0.481481 | 0.566667 | 0.714286 | 0.319672  | 0.342105 | 0.258065 | 0.205128 |  4.0   |  1.0   |     0      |     0     |      1      |     1      |     1    |    1   |    0    |      0     |   0    |    1    |     1      |
| 0.878378 |  0.00  | 0.206009  | 0.751634 | 0.604938 | 0.533333 | 0.571429 | 0.475410  | 0.500000 | 0.879032 | 0.435897 |  4.0   |  0.0   |     1      |     1     |      0      |     0      |     1    |    1   |    0    |      1     |   1    |    0    |     1      |
| 0.878378 |  0.25  | 0.639485  | 0.470588 | 0.395062 | 0.433333 | 0.428571 | 0.393443  | 0.447368 | 0.104839 | 0.256410 |  3.0   |  0.0   |     1      |     1     |      1      |     1      |     1    |    1   |    1    |      1     |   1    |    1    |     1      |
| 0.716216 |  0.50  | 1.000000  | 0.163399 | 0.111111 | 0.066667 | 0.171429 | 0.393443  | 0.500000 | 0.532258 | 0.435897 |  1.0   |  0.0   |     0      |     1     |      1      |     1      |     0    |    1   |    1    |      0     |   0    |    0    |     0      |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate LLM hallucinations in data conversions, ensure clean, well-structured input data and provide explicit instructions. Test incrementally on smaller datasets to identify errors early. Use cross-validation with multiple methods and incorporate human review to verify accuracy. Supplement conversions with dedicated data transformation tools for reliability. Implement a feedback loop to refine outputs and maintain logs to track inconsistencies. Error reporting helps identify patterns and improve future results. By combining these strategies, you can enhance accuracy, minimize hallucinations, and ensure reliable data conversions for downstream applications.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [22]:
# Code to rename all the columns in the dataset
column_mapping = {
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia'
}

dataset_final.rename(columns=column_mapping, inplace=True)

# Print the updated DataFrame
print(dataset_final)

          age  blood pressure  blood glucose random  blood urea  \
0    0.743243            0.50              0.442060    0.901961   
1    0.770270            1.00              0.901288    0.163399   
2    0.905405            0.50              0.785408    0.862745   
3    0.202703            0.75              0.158798    0.196078   
4    0.783784            1.00              0.399142    0.287582   
5    0.729730            0.00              0.935622    0.169935   
6    0.675676            0.75              0.253219    0.633987   
7    0.851351            0.25              0.618026    0.562092   
8    0.540541            0.00              0.399142    0.535948   
9    0.756757            0.25              0.223176    0.209150   
10   0.662162            0.50              0.618026    0.411765   
11   0.783784            0.00              0.725322    0.313725   
12   0.878378            0.00              0.206009    0.751634   
13   0.878378            0.25              0.639485    0.47058

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [32]:
# Converted SQL to Pandas code

# Group by 'Target' and count occurrences
result = dataset_normalized.groupby('Target_ckd').size().reset_index(name='count')

# Print the result
print(result)

   Target_ckd  count
0           0    111
1           1     23


*Describe what the above code does here*

The code counts the entries of the target column which implies how many patients are diagnosed with chronic kidney disease(indicated by 1) and the number of patients that are not(indicated by 0).